File: mm-ts/src/core/binder.ts
import { NodeScalar, NodeObject, NodeArray, NodeNull } from '../ir/ast';
import { Tag } from '../ir/tag';
import { Node } from '../ir/ast';
import { ValueType } from '../ir/value-type';

export type Constructor<T> = new () => T;

export class Binder {
  bind<T>(node: Node, type: Constructor<T> | T): T {
    const result =
      typeof type === 'function' ? new (type as Constructor<T>)() : type;
    const { value, error } = this.bindNode(node, result);
    if (error) {
      throw new Error(error);
    }
    return value;
  }

  bindNode(node: Node, target: any): { value: any; error: string | null } {
    // Handle NodeNull - return null value
    if (node instanceof NodeNull) {
      return { value: null, error: null };
    }

    if (node instanceof NodeScalar) {
      return { value: node.getValue(), error: null };
    }

    // When target is null/undefined, create dynamic containers
    // This mirrors Go's interface{} handling in Bind
    if (target === null || target === undefined) {
      if (node instanceof NodeObject) {
        const properties = node.getProperties();
        const result: Record<string, any> = {};
        for (const [key, valueNode] of Object.entries(properties)) {
          const { value, error } = this.bindNode(valueNode, undefined);
          if (error) {
            return {
              value: result,
              error: `failed to bind field ${key}: ${error}`,
            };
          }
          result[key] = value;
        }
        return { value: result, error: null };
      }

      if (node instanceof NodeArray) {
        const elements = node.getElements();
        const result: any[] = [];
        for (let i = 0; i < elements.length; i++) {
          const el = elements[i]!;
          const { value, error } = this.bindNode(el, undefined);
          if (error) {
            return {
              value: result,
              error: `failed to bind array item ${i}: ${error}`,
            };
          }
          result.push(value);
        }
        return { value: result, error: null };
      }
    }

    if (node instanceof NodeObject) {
      const tag = node.getTag();
      if (tag?.type === ValueType.Obj) {
        return this.convertObj(node, target);
      } else {
        return this.convertMap(node, target);
      }
    }

    if (node instanceof NodeArray) {
      const tag = node.getTag();
      if (tag?.type === ValueType.Arr) {
        return this.convertArr(node, target);
      } else {
        return this.convertVec(node, target);
      }
    }

    return { value: target, error: `unsupported node type: ${typeof node}` };
  }

  private convertObj(
    obj: NodeObject,
    target: any,
  ): { value: any; error: string | null } {
    const tag = obj.getTag();
    if (tag?.nullable && target === null) {
      return {
        value: target,
        error: 'convertObj requires object type, got null',
      };
    }
    if (typeof target !== 'object' || target === null) {
      return {
        value: target,
        error: `convertObj requires object type, got ${typeof target}`,
      };
    }

    const properties = obj.getProperties();
    for (const [key, valueNode] of Object.entries(properties)) {
      const camelKey = key.replace(/_([a-z])/g, (_, c) => c.toUpperCase());
      if (!(camelKey in target)) {
        // For plain objects (not class instances), create fields dynamically
        // This mirrors Go's interface{} handling in Bind
        if (
          target.constructor === Object ||
          target.constructor === undefined
        ) {
          const { value, error } = this.bindNode(valueNode, undefined);
          if (error) {
            return {
              value: target,
              error: `failed to bind field ${camelKey}: ${error}`,
            };
          }
          target[camelKey] = value;
          continue;
        }
        return { value: target, error: `struct has no field '${camelKey}'` };
      }

      const { value, error } = this.bindNode(valueNode, target[camelKey]);
      if (error) {
        return {
          value: target,
          error: `failed to bind field ${camelKey}: ${error}`,
        };
      }

      target[camelKey] = value;
    }

    return { value: target, error: null };
  }

  private convertMap(
    obj: NodeObject,
    target: any,
  ): { value: any; error: string | null } {
    if (
      typeof target !== 'object' ||
      target === null ||
      Array.isArray(target)
    ) {
      return { value: target, error: `convertMap requires object` };
    }

    const properties = obj.getProperties();
    for (const [key, valueNode] of Object.entries(properties)) {
      const { value, error } = this.bindNode(valueNode, target[key]);
      if (error) {
        return { value: target, error: `field ${key}: ${error}` };
      }
      target[key] = value;
    }
    return { value: target, error: null };
  }

  private convertArr(
    arr: NodeArray,
    target: any,
  ): { value: any; error: string | null } {
    if (!Array.isArray(target)) {
      return { value: target, error: `convertArr requires array` };
    }

    const elements = arr.getElements();
    const size = arr.getTag()?.size || elements.length;

    if (target.length !== size) {
      return {
        value: target,
        error: `array length mismatch: expected ${size}, got ${target.length}`,
      };
    }

    for (let i = 0; i < elements.length; i++) {
      const el = elements[i];
      if (!el) continue;

      const { value, error } = this.bindNode(el, target[i]);
      if (error) {
        return { value: target, error: `array[${i}]: ${error}` };
      }
      target[i] = value;
    }

    return { value: target, error: null };
  }

  private convertVec(
    arr: NodeArray,
    target: any,
  ): { value: any; error: string | null } {
    if (!Array.isArray(target)) {
      // Create a new array for dynamic binding
      target = [];
    }

    const elements = arr.getElements();
    target.length = 0;

    for (const el of elements) {
      const defaultVal = this.createDefaultValue(el.getTag());
      const { value, error } = this.bindNode(el, defaultVal);
      if (error) {
        return { value: target, error };
      }
      target.push(value);
    }

    return { value: target, error: null };
  }

  private createDefaultValue(tag: Tag | null): any {
    if (!tag) return null;

    switch (tag.type) {
      case ValueType.I:
      case ValueType.U:
      case ValueType.I64:
      case ValueType.U64:
      case ValueType.Bigint:
        return 0n;

      case ValueType.I8:
      case ValueType.I16:
      case ValueType.I32:
      case ValueType.U8:
      case ValueType.U16:
      case ValueType.U32:
        return 0;

      case ValueType.F32:
      case ValueType.F64:
        return 0.0;

      case ValueType.Str:
      case ValueType.Email:
      case ValueType.Url:
      case ValueType.Ip:
      case ValueType.Uuid:
      case ValueType.Decimal:
      case ValueType.Enums:
        return '';

      case ValueType.Decimal:
        return '0.0';

      case ValueType.Bool:
        return false;

      case ValueType.Bytes:
      case ValueType.Media:
        return new Uint8Array();

      case ValueType.Datetime:
      case ValueType.Date:
      case ValueType.Time:
        return new Date(0);

      default:
        return null;
    }
  }
}

export function bindJSONC<T>(node: Node, Type: new () => T): T {
  const binder = new Binder();
  return binder.bind(node, Type);
}

File: mm-kt/src/main/kotlin/io/github/metamessage/core/Binder.kt
package io.github.metamessage.core

import io.github.metamessage.ir.Node
import io.github.metamessage.ir.NodeArray
import io.github.metamessage.ir.NodeNull
import io.github.metamessage.ir.NodeObject
import io.github.metamessage.ir.NodeScalar
import io.github.metamessage.ir.Tag
import io.github.metamessage.ir.ValueType
import java.math.BigInteger
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.util.UUID

object Binder {
    @Suppress("UNCHECKED_CAST")
    fun <T> bind(node: Node, clazz: Class<T>): T {
        return bind(node, clazz, null)
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> bind(node: Node, clazz: Class<T>, elementClass: Class<*>? = null): T {
        when (node) {
            is NodeNull -> {
                @Suppress("UNCHECKED_CAST") return null as T
            }
            is NodeObject -> {
                val tag = node.tag
                if (tag != null && tag.type == ValueType.OBJ) {
                    val inst = clazz.getDeclaredConstructor().newInstance()
                    convertObj(node, inst as Any)
                    return inst
                } else {
                    @Suppress("UNCHECKED_CAST") val inst = mutableMapOf<String, Any?>()
                    convertMap(node, inst)
                    return inst as T
                }
            }
            is NodeArray -> {
                val tag = node.tag
                if (tag != null && tag.size > 0 && tag.type == ValueType.ARR) {
                    @Suppress("UNCHECKED_CAST") return convertArr(node, elementClass ?: clazz) as T
                } else {
                    if (elementClass != null) {
                        @Suppress("UNCHECKED_CAST") return convertVec(node, elementClass) as T
                    }
                    @Suppress("UNCHECKED_CAST") return convertVecDynamic(node) as T
                }
            }
            is NodeScalar -> {
                return convertScalar(node, clazz)
            }
            else ->
                    throw IllegalArgumentException(
                            "unsupported node type: ${node::class.java.name}"
                    )
        }
    }

    private fun convertObj(obj: NodeObject, out: Any) {
        val outClazz = out.javaClass
        val nameToField = mutableMapOf<String, java.lang.reflect.Field>()
        for (f in outClazz.declaredFields) {
            if (java.lang.reflect.Modifier.isStatic(f.modifiers)) continue
            nameToField[f.name] = f
        }

        for (field in obj.fields) {
            val fieldKey = field.key
            val camelName = SnakeToCamel.convert(fieldKey)
            val structField = nameToField[camelName] ?: continue
            structField.isAccessible = true
            try {
                val fieldVal = materialize(structField, field.value)
                structField.set(out, fieldVal)
            } catch (e: Exception) {
                throw RuntimeException("failed to bind field ${field.key}: ${e.message}", e)
            }
        }
    }

    private fun convertMap(obj: NodeObject, out: MutableMap<String, Any?>) {
        for (field in obj.fields) {
            val key = field.key
            val value =
                    when (val v = field.value) {
                        is NodeNull -> null
                        is NodeScalar -> convertScalarToAny(v)
                        is NodeObject -> {
                            val map = mutableMapOf<String, Any?>()
                            convertMap(v, map)
                            map
                        }
                        is NodeArray -> {
                            convertVecDynamic(v)
                        }
                        else -> null
                    }
            out[key] = value
        }
    }

    private fun convertArr(arr: NodeArray, elementClass: Class<*>): Any {
        val list = mutableListOf<Any?>()
        for (item in arr.items) {
            when (item) {
                is NodeScalar -> list.add(convertScalarToAny(item))
                is NodeObject -> {
                    val inst = elementClass.getDeclaredConstructor().newInstance()
                    convertObj(item, inst as Any)
                    list.add(inst)
                }
                else -> list.add(null)
            }
        }
        return list
    }

    @Suppress("UNCHECKED_CAST")
    private fun convertVec(arr: NodeArray, field: java.lang.reflect.Field): Any {
        val elementClass = extractElementType(field)
        return convertVec(arr, elementClass)
    }

    @Suppress("UNCHECKED_CAST")
    private fun convertVec(arr: NodeArray, elementClass: Class<*>): Any {
        val list = mutableListOf<Any?>()
        for (item in arr.items) {
            when (item) {
                is NodeScalar -> list.add(convertScalarToAny(item))
                is NodeObject -> {
                    val inst = elementClass.getDeclaredConstructor().newInstance()
                    convertObj(item, inst as Any)
                    list.add(inst)
                }
                else -> list.add(null)
            }
        }
        return list
    }

    private fun extractElementType(field: java.lang.reflect.Field): Class<*> {
        val genericType = field.genericType
        if (genericType is java.lang.reflect.ParameterizedType) {
            val typeArgs = genericType.actualTypeArguments
            if (typeArgs.isNotEmpty()) {
                val arg = typeArgs[typeArgs.size - 1]
                if (arg is Class<*>) {
                    return arg
                }
                if (arg is java.lang.reflect.ParameterizedType) {
                    val rawType = arg.rawType
                    if (rawType is Class<*>) {
                        return rawType
                    }
                }
            }
        }
        return Any::class.java
    }

    private fun convertVecDynamic(arr: NodeArray): List<Any?> {
        val list = mutableListOf<Any?>()
        for (item in arr.items) {
            when (item) {
                is NodeScalar -> list.add(convertScalarToAny(item))
                is NodeObject -> {
                    val map = mutableMapOf<String, Any?>()
                    convertMap(item, map)
                    list.add(map)
                }
                else -> list.add(null)
            }
        }
        return list
    }

    @Suppress("UNCHECKED_CAST", "UNUSED_PARAMETER")
    private fun <T> convertScalar(value: NodeScalar, clazz: Class<T>): T {
        val tag = value.tag ?: Tag.empty()
        val data = value.data
        val text = value.text

        if (tag.isNull) {
            @Suppress("UNCHECKED_CAST") return null as T
        }

        return when (tag.type) {
            ValueType.DATETIME -> {
                (data as? LocalDateTime ?: LocalDateTime.of(1970, 1, 1, 0, 0, 0)) as T
            }
            ValueType.DATE -> {
                when (data) {
                    is LocalDate -> data as T
                    is LocalDateTime ->
                            (if (clazz == LocalDate::class.java) data.toLocalDate() else data) as T
                    else -> LocalDate.of(1970, 1, 1) as T
                }
            }
            ValueType.TIME -> {
                when (data) {
                    is LocalTime -> data as T
                    is LocalDateTime ->
                            (if (clazz == LocalTime::class.java) data.toLocalTime() else data) as T
                    else -> LocalTime.of(0, 0, 0) as T
                }
            }
            ValueType.BIGINT -> {
                (data as? BigInteger ?: BigInteger.ZERO) as T
            }
            ValueType.UUID -> {
                (data as? UUID ?: UUID(0, 0)).toString() as T
            }
            ValueType.DECIMAL, ValueType.EMAIL -> {
                (data as? String ?: text) as T
            }
            ValueType.IP -> {
                when (clazz) {
                    String::class.java ->
                            ((data as? java.net.InetAddress)?.hostAddress ?: text) as T
                    else -> (data as? java.net.InetAddress) as T
                }
            }
            ValueType.URL -> {
                when (clazz) {
                    String::class.java -> ((data as? java.net.URI)?.toString() ?: text) as T
                    else -> (data as? java.net.URI) as T
                }
            }
            ValueType.ENUMS -> {
                text as T
            }
            ValueType.I -> {
                val num = (data as? Number) ?: text.toInt()
                convertNumberToType(num, clazz) as T
            }
            ValueType.I8 -> {
                val num = (data as? Number) ?: text.toByte()
                convertNumberToType(num, clazz) as T
            }
            ValueType.I16 -> {
                val num = (data as? Number) ?: text.toShort()
                convertNumberToType(num, clazz) as T
            }
            ValueType.I32 -> {
                val num = (data as? Number) ?: text.toInt()
                convertNumberToType(num, clazz) as T
            }
            ValueType.I64 -> {
                val num = (data as? Number) ?: text.toLong()
                convertNumberToType(num, clazz) as T
            }
            ValueType.U -> {
                val num = (data as? Number) ?: text.toInt()
                convertNumberToType(num, clazz) as T
            }
            ValueType.U8 -> {
                val num = (data as? Number) ?: text.toShort()
                convertNumberToType(num, clazz) as T
            }
            ValueType.U16 -> {
                val num = (data as? Number) ?: text.toInt()
                convertNumberToType(num, clazz) as T
            }
            ValueType.U32 -> {
                val num = (data as? Number) ?: text.toInt()
                convertNumberToType(num, clazz) as T
            }
            ValueType.U64 -> {
                val num = (data as? Number) ?: text.toLong()
                convertNumberToType(num, clazz) as T
            }
            ValueType.F32 -> {
                val num = (data as? Number) ?: text.toFloat()
                convertNumberToType(num, clazz) as T
            }
            ValueType.F64 -> {
                val num = (data as? Number) ?: text.toDouble()
                convertNumberToType(num, clazz) as T
            }
            ValueType.STR -> {
                (data as? String ?: text) as T
            }
            ValueType.BOOL -> {
                (data as? Boolean ?: text.toBoolean()) as T
            }
            ValueType.BYTES -> {
                (data as? ByteArray) as T
            }
            else -> throw IllegalArgumentException("unsupported type: ${tag.type}")
        }
    }

    private fun convertScalarToAny(value: NodeScalar): Any? {
        return convertScalar(value, Any::class.java)
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> convertNumberToType(num: Number, clazz: Class<T>): Any {
        return when (clazz) {
            Byte::class.java, java.lang.Byte.TYPE -> num.toByte()
            Short::class.java, java.lang.Short.TYPE -> num.toShort()
            Int::class.java, java.lang.Integer.TYPE -> num.toInt()
            Long::class.java, java.lang.Long.TYPE -> num.toLong()
            Float::class.java, java.lang.Float.TYPE -> num.toFloat()
            Double::class.java, java.lang.Double.TYPE -> num.toDouble()
            BigInteger::class.java -> BigInteger.valueOf(num.toLong())
            else -> num
        } as
                Any
    }

    private fun materialize(f: java.lang.reflect.Field, node: Node): Any? {
        return when (node) {
            is NodeNull -> null
            is NodeScalar -> convertScalar(node, f.type)
            is NodeObject -> {
                val tag = node.tag
                if (tag != null && tag.type == ValueType.MAP) {
                    val map = mutableMapOf<String, Any?>()
                    convertMap(node, map)
                    map
                } else {
                    val inst = f.type.getDeclaredConstructor().newInstance()
                    convertObj(node, inst)
                    inst
                }
            }
            is NodeArray -> {
                convertVec(node, f)
            }
            else -> null
        }
    }
}
